src/docfx/lib/markdown/validation/DocsValidationExtension.cs

```diff
@@ -88,9 +53,93 @@ internal static class DocsValidationExtension
                 {
                     contentValidator.ValidateHeadings(inclusion, inclusionNodes, true);
                 }
+
+                foreach (var codeBlock in codeBlockItemList)
+                {
+                    contentValidator.ValidateCodeBlock(currentFile, codeBlock.codeBlockItem, codeBlock.IsIncluded);
+                }
             });
         }
 
+        private static void BuildHeadingNodes(
+            MarkdownObject node,
+            MarkdownVisitContext context,
+            MarkdownEngine markdownEngine,
+            List<ContentNode> documentNodes,
+            Dictionary<Document, List<ContentNode>> inclusionDocumentNodes,
+            bool? isCanonicalVersion)
+        {
+            ContentNode? documentNode = null;
+            if (node is HeadingBlock headingBlock)
+            {
+                var headingNode = CreateValidationNode<Heading>(context, isCanonicalVersion);
+
+                headingNode.Level = headingBlock.Level;
+                headingNode.SourceInfo = headingBlock.GetSourceInfo();
+                headingNode.Content = GetHeadingContent(headingBlock); // used for reporting
+                headingNode.HeadingChar = headingBlock.HeaderChar;
+                headingNode.RenderedPlainText = markdownEngine.ToPlainText(headingBlock); // used for validation
+                headingNode.IsVisible = MarkdigUtility.IsVisible(headingBlock);
+
+                documentNode = headingNode;
+            }
+            else if (node is LeafBlock leafBlock)
+            {
+                var contentNode = CreateValidationNode<ContentNode>(context, isCanonicalVersion);
+
+                contentNode.SourceInfo = node.GetSourceInfo();
```

Comment:
put these two into CreateValidationNode

Reply:
Got it.
SourceInfo is ok,  but IsVisible belongs to ContentNode.